x/gamm/keeper/multihop.go

```diff
@@ -16,12 +16,55 @@ func (k Keeper) MultihopSwapExactAmountIn(
 	tokenIn sdk.Coin,
 	tokenOutMinAmount sdk.Int,
 ) (tokenOutAmount sdk.Int, err error) {
-	for i, route := range routes {
-		swapFeeMultiplier := sdk.OneDec()
-		if types.SwapAmountInRoutes(routes).IsOsmoRoutedMultihop() {
-			swapFeeMultiplier = types.MultihopSwapFeeMultiplierForOsmoPools.Clone()
+	var route0Incentivized bool
+	var route1Incentivized bool
+	additiveSwapFee := sdk.ZeroDec()
+	maxSwapFee := sdk.ZeroDec()
+
+	// get list of all incentivized pools
+	incentivizedPools := k.poolIncentivesKeeper.GetAllIncentivizedPools(ctx)
+
+	// in this loop, we check if:
+	// - the route is of length 2
+	// - route 1 and route 2 don't trade via the same pool
+	// - route 1 contains uosmo
+	// - both route 1 and route 2 are incentivized pools
+	// if all of the above is true, then we collect the additive and max fee between the two pools to later calculate the following:
+	// total_swap_fee = total_swap_fee = max(swapfee1, swapfee2)
+	// fee_per_pool = total_swap_fee * ((pool_fee) / (swapfee1 + swapfee2))
+	if types.SwapAmountInRoutes(routes).IsOsmoRoutedMultihop() {
+		for _, route := range routes {
+			// check both pools being router to have internal incentives
+			for _, pool := range incentivizedPools {
+				if routes[0].PoolId == pool.PoolId {
+					route0Incentivized = true
+				}
+				if routes[1].PoolId == pool.PoolId {
+					route1Incentivized = true
+				}
+				if route0Incentivized && route1Incentivized {
+					break
+				}
+			}
+			pool, poolErr := k.getPoolForSwap(ctx, route.PoolId)
+			if poolErr != nil {
+				return sdk.Int{}, poolErr
+			}
+			swapFee := pool.GetSwapFee(ctx)
+			// add to the swap fee accumulator
+			additiveSwapFee = additiveSwapFee.Add(swapFee)
+			// determine the max swap fee
+			if swapFee.GT(maxSwapFee) {
+				maxSwapFee = swapFee
+			}
 		}
+	}
+
+	if additiveSwapFee.Quo(sdk.NewDec(2)).GT(maxSwapFee) {
+		maxSwapFee = additiveSwapFee.Quo(sdk.NewDec(2))
+	}
```

Comment:
We should move this to its own function

Reply:
moved it to its own func, sadly cant have too much reuse between SwapOut and SwapIn due to having different route types